### ISSUE TITLE
Possible regression in dev mode in v16.3

### ISSUE THREAD

**Issue by neoziro** [facebook/react#12502]
[This example](https://codesandbox.io/s/1zn2o513pq) presents a huge performance drop between React 16.2 and React 16.3.

In investigated it and noticed several things:

- Perfomance drop is only visible using development bundle
- This is not related to styled-components (a simple div presents the same problem)
- I tried to investigate in performance tab of Chrome dev tools and I noticed that I get React measures even without "?react_perf"
- React Tree Reconciliation seems very long, do not know why

It is not a big problem for me, but it looks weird, so I decided to [share it on Twitter](https://twitter.com/neoziro/status/979756579296817154),  then @gaearon [invited me to submit an issue about it](https://twitter.com/dan_abramov/status/980036013518139398).

**Comment by gaearon** [comment]
Can you turn this into a smaller self-contained example without third party libs?

**Comment by neoziro** [comment]
I will try to do it but right now I don’t have time, I think doing a setState every 16ms on 20 components in parallel should be sufficient to make it slow.

**Comment by sebmarkbage** [comment]
Note that if you’re using the StrictMode component, this is expected since it double renders. If you’re not, then it is not expected and probably a bug.

**Comment by alshakero** [comment]
Here is a fairly simple comparision. Exactly the same code with different react version. 

- 16.3: https://codesandbox.io/s/px9ny4v50 | 201.82ms per frame
- 16.2: https://codesandbox.io/s/m369j7jomp | 54.76ms per frame

**Comment by mlazowik** [comment]
* 16.2: <img width="1281" alt="16 2" src="https://user-images.githubusercontent.com/1978721/38168349-c402a1fc-3549-11e8-8ac8-0e976a916943.png">
 * 16.3: <img width="1118" alt="16 3" src="https://user-images.githubusercontent.com/1978721/38168350-c6089b5a-3549-11e8-9b3f-020b4b28f62e.png">

Looks like scheduling is somehow off and leaves idle gaps?

**Comment by gaearon** [comment]
We're spending *quite* a bit more time in `performUnitOfWork` (which is the hottest path).
I think [this](https://github.com/facebook/react/blob/fa8e67893fca1b3902637129972032bca248a584/packages/react-reconciler/src/ReactFiberScheduler.js#L849) might be the culprit.

Want to try commenting that line out in the bundle and see if it helps?

**Comment by mlazowik** [comment]
@gaearon yep, this is it.

**Comment by gaearon** [comment]
Good. I wonder why. Is it the allocation, the enumeration over a fiber, or changing hidden class of the target itself.

Can you try

```js
Object.assign(stashedWorkInProgressProperties, workInProgress);
```

instead and see if it’s still as bad?

**Comment by mlazowik** [comment]
`stashedWorkInProgressProperties` is undefined there, so this errors out.

**Comment by mlazowik** [comment]
@gaearon 

```js
if (true && replayFailedUnitOfWorkWithInvokeGuardedCallback) {
    allocStashed();
    copyWIP();
}

function allocStashed() {
    stashedWorkInProgressProperties = {};
}

function copyWIP() {
    Object.assign(stashedWorkInProgressProperties, workInProgress);
}
```

<img width="1095" alt="zrzut ekranu 2018-04-01 o 02 02 49" src="https://user-images.githubusercontent.com/1978721/38168584-d458f1a8-3550-11e8-9239-1f6e33f13602.png">

**Comment by gaearon** [comment]
Seems like it's still quite a lot. I bet we need to preserve the hidden class and explicitly write out all assignments.

**Comment by mlazowik** [comment]
Hm, maybe you could write a generic function to do that based on type annotations?

**Comment by mlazowik** [comment]
@gaearon an ad-hoc confirmation that hidden class is the issue:

```js
if (true && replayFailedUnitOfWorkWithInvokeGuardedCallback) {
    stashedWorkInProgressProperties = createFiber(
        workInProgress.tag,
        workInProgress.pendingProps,
        workInProgress.key,
        workInProgress.mode
    );
}
```

<img width="849" alt="zrzut ekranu 2018-04-01 o 02 39 15" src="https://user-images.githubusercontent.com/1978721/38168737-19ed7522-3556-11e8-9e25-0f825d537172.png">

**Comment by mlazowik** [comment]
I forgot to actually copy.
```js
if (true && replayFailedUnitOfWorkWithInvokeGuardedCallback) {
    stashedWorkInProgressProperties = createFiber(
        workInProgress.tag,
        workInProgress.pendingProps,
        workInProgress.key,
        workInProgress.mode
    );
    Object.assign(stashedWorkInProgressProperties, workInProgress);
}
```

This is still slow, but faster than with `stashedWorkInProgressProperties = {}`.

**Comment by gaearon** [comment]
I tried locally and handwritten version seems pretty fast.

**Comment by gaearon** [comment]
https://github.com/facebook/react/pull/12510

**Comment by neoziro** [comment]
Tested on the original example and it is much better, thanks!

**Comment by gaearon** [comment]
Fixed in 16.3.1.
